cpu/kinetis/periph/gpio.c

```diff
@@ -308,11 +308,18 @@ static inline void irq_handler(PORT_Type *port, int port_num)
     /* take interrupt flags only from pins which interrupt is enabled */
     uint32_t status = port->ISFR;
 
-    for (int i = 0; i < 32; i++) {
-        if ((status & (1u << i)) && (port->PCR[i] & PORT_PCR_IRQC_MASK)) {
-            port->ISFR = (1u << i);
-            int ctx = get_ctx(port_num, i);
-            isr_ctx[ctx].cb(isr_ctx[ctx].arg);
+    unsigned pin = 0;
+    while (status) {
+        unsigned tz = __builtin_ctz(status);
+        pin += tz;
```

Comment:
What's in status? A bitfield with the pending ISRs I suppose?

If yes, I think we could simplify this, as "ctz" should directly return the number of the LSB (and thus pin) starting from zero:

```
while(status) {
  /* get position of first bit set in status */
  unsigned pin = __builtin_ctz(status);
 /* clear it */
  status &= ~(1<<pin);
  if (...
```

I wouldn't bother, as the compiler usually figures this itself, but it took me a bit to understand the current version.

Reply:
btw, did you try ```bitarithm_lsb()```?

Reply:
The status variable is a bit field where 1 means the pin interrupt has occurred, 0 means no interrupt has occurred.
I didn't test bitarithm_lsb yet. I will make a quick test now while I have the measurement set up still connected.

Reply:
All measurements were performed with the logic analyzer set to 100 MHz sample rate.

Tested with bitarithm_lsb on CM0+ (frdm-kw41z). The timing is on average 5.7µs with bitarithm_lsb on CM0+ (frdm-kw41z), `BITARITHM_LSB_LOOKUP` is defined, `BITARITHM_LSB_BUILTIN` is not defined.
There is quite a bit of other things adding to the measured delay though, so if you want to perform a direct comparison of the MultiplyDeBruijnBitPosition implementation against the GCC builtin armv6 asm implementation I think we will need a different test program.

One more measurement was taken as well, I moved the LED0_ON macro to the first line of the GPIO interrupt callback routine (the one passed as an argument to gpio_init_int). These measurements seem to confirm the improvement from builtin_ctz to bitarithm_lsb

Average time from GPIO pin flank to first line of `cb`:
With `bitarithm_lsb`: 4.4 µs
With `__builtin_ctz`: 4.8 µs

Reply:
Thanks for the suggestion on the bit clearing, it saved another 0.3-0.4 µs on average.

Reply:
Looking at the generated assembly, your suggestion makes the compiler generate a single `bics` instruction (bit clear), while the other version needs some adds and lsrs instructions

Reply:
> if you want to perform a direct comparison of the MultiplyDeBruijnBitPosition implementation against the GCC builtin armv6 asm implementation I think we will need a different test program.

I did some testing a while ago using "tests/bitarithm_timing" and the scheduler benchmarks.

IIRC the "naive" version (```i= 0; while (!val&1) { i++; val>>=1; return ;}```, and also the GCC builtin without clz) where not constant in time regarding to input. E.g., "000001" was faster than "1000000". "MultiplyDeBruijnBitPosition" is constant, but a tad slower on average. You don't have an easy way of testing the delays for different bits, e.g., 31 and 0? :)
Using "MultiplyDeBruijnBitPosition" was having quite some positive impact on context switch performance, that's why it is default for ARM's without CLZ.

For ARM with CLZ, the builtin is very fast and constant in all cases.

Reply:
Not all pins are usable since they are already used for other functions, but we might find some pin numbered 14-19 which may be usable for a GPIO IRQ. I'll try to test something later tonight or in the morning.

Reply:
Should be, maybe I forgot to push. Will update

Reply:
pushed